mtp_api/apps/security/serializers.py

```diff
@@ -349,3 +352,76 @@ def reject(self, by):
             raise ValidationError(
                 detail=e.message_dict,
             )
+
+
+class CheckAutoAcceptRuleStateSerializer(serializers.ModelSerializer):
+    active = serializers.BooleanField(required=False)
+    added_by = BasicUserSerializer(read_only=True)
+
+    class Meta:
+        model = CheckAutoAcceptRuleState
+        fields = [
+            'added_by',
+            'active',
+            'reason',
+            'created'
+        ]
+        read_only_fields = (
+            'id',
+        )
+
+
+class CheckAutoAcceptRuleSerializer(serializers.ModelSerializer):
+    states = CheckAutoAcceptRuleStateSerializer(many=True, required=True)
+
+    def validate(self, attrs):
+        if len(attrs['states']) != 1:
+            raise ValidationError(
+                f'When creating or updating an auto-accept rule, states must be of length 1, not {len(attrs.states)}'
+            )
```

Comment:
Interesting, so you changed the serializer and it's now getting a `states` array rather than a `reason`.

So if I understand this correctly you got rid of the abstraction and now don't try to hide the rule/states implementation details but this means the client needs to pass `states` explicitly.

So:
1. on `create`, `states` would have the initial state. This also means that potentially the client can create a non-active state
2. on `update`, `states` only has the new state which is added

Is that correct?

Reply:
Yeah, so I had originally intended on keeping the abstraction, but Django Rest Framework Serializers don't really facilitate expecting a differing data structure on `create`, `update` and serialization, at least not in a nice way, so I moved away from that. I'm happy to make the endpoint structure more explicit with respect to what's happening underneath (i.e. removing the `UpdateModelMixin` from `CheckAutoAcceptRuleView`, and creating a new `AutoAcceptRuleState` endpoint that's create only and has the `CheckAutoAcceptRule.id` as a positional URL parameter, and then removing `CheckAutoAcceptSerializer.update` in favour of the default `CheckAutoAcceptState.create` implementation from `ModelSerializer`. That's probably a more RESTful way of approaching the problem.

Reply:
Ah, I remember why I didn't do this originally; there's not really a nice way to ensure that `CheckAutoAcceptRuleSerializer.create` tolerates the addition of the `reason` parameter for the writable case; for serializing it's simple enough to do this using `serializers.SerializerMethodFields` and a `get_*` method for the read case, but no equivalent exists for the write case. I really don't want to break the creation up into API calls into multiple API calls as that's just asking for data integrity problems.

Reply:
yeah i agree. it's easiest to just keep it all together; the logic is not complex